### DISCUSSION TITLE
multi-scale tessellation

### DISCUSSION THREAD

**Discussion by Ming-is** [neperfepx/neper#401] [Q&A] [answered]
I want to generate a mosaic model with multi-scale tessellation. For example, 

1.I want to generate a mosaic model of 2-scale tessellation. For example, 
![image](https://user-images.githubusercontent.com/71123582/156878667-a0ec7284-4401-46d4-b276-2b0dbcead9f8.png)

to generate a 10 grain “gg” morphology model, specify that the fifth grain is subdivided into a layered structure with a width of 0.1, and the other grains are not subdivided. What instructions should I use?

2..I want to generate a mosaic model of 3-scale tessellation.For example，to generate a 10 grain “gg”morphology model, specify that the fifth grain is subdivided into five clusters, and the clusters are subdivided into a layered structure with a width of 0.1， and the other grains are not subdivided. What instructions should I use?

**Comment by rquey** [comment, marked as answer]
1:

You have to use a [multiscale cell file](https://neper.info/doc/fileformat.html#multiscale-cell-file) to pass different option values to each first-scale cell:

```
neper -T -n "10::msfile(n_msfile)" -morpho "gg::msfile(morpho_msfile)" -o d
```

where the multiscale files are:

n_msfile:
```
1 1
2 1
3 1
4 1
5 from_morpho
6 1
7 1
8 1
9 1
10 1
```

morpho_msfile:
```
1 voronoi
2 voronoi
3 voronoi
4 voronoi
5 lamellar(w=0.1)
6 voronoi
7 voronoi
8 voronoi
9 voronoi
10 voronoi
```

Visualize:

```
neper -V d.tess -datacellcol "scaleid(1)" -datacelltrs 0.5 -print img
```
![image](https://user-images.githubusercontent.com/2659376/157269047-ef0f0208-c101-4e00-9f39-1bee1ebe4b8e.png)

2:

Same principle:

```
neper -T -n "10::msfile(n_msfile2)::msfile(n_msfile3)" -morpho "gg::voronoi::msfile(morpho_msfile3)" -o d
```

where the multiscale files are:

n_msfile2:
```
1 1
2 1
3 1
4 1
5 5
6 1
7 1
8 1
9 1
10 1
```

n_msfile3:
```
1::1 1
2::1 1
3::1 1
4::1 1
5::1 from_morpho
5::2 from_morpho
5::3 from_morpho
5::4 from_morpho
5::5 from_morpho
6::1 1
7::1 1
8::1 1
9::1 1
10::1 1
```

morpho_msfile3:
```
1::1 voronoi
2::1 voronoi
3::1 voronoi
4::1 voronoi
5::1 lamellar(w=0.1)
5::2 lamellar(w=0.1)
5::3 lamellar(w=0.1)
5::4 lamellar(w=0.1)
5::5 lamellar(w=0.1)
6::1 voronoi
7::1 voronoi
8::1 voronoi
9::1 voronoi
10::1 voronoi
```

Visualize:
```
neper -V d.tess -datacellcol "scaleid(1)" -datacelltrs 0.5 -print img
```

![image](https://user-images.githubusercontent.com/2659376/157270057-c52638a1-5bf7-43c5-abb0-016e0157e66c.png)

It is also possible to merge the different multiscale cell files (`n_msfile2` and `n_msfile3`, in the above) - Neper scans the file until it finds the correct multiscale id at the beginning of the line.

**Reply by Ming-is** [reply]
Thank you for your perfect reply. If so, can neper further specify the custom grain orientation for the daughter cells, for example, generate a 5-grain mosaic, the 5th grain is divided into 5 daughter cells, other grains use random orientation, and the daughter cells use the custom orientation file to define the orientation. What instructions should be used？

**Reply by rquey** [reply]
Yes, it works in a similar way (`-morpho` is only to produce a nicer tessellation):

```
neper -T -n "5::msfile(n_msfile)" -ori "random::msfile(ori_msfile)" -morpho diameq:1,sphericity:0.85::diameq:1,sphericity:0.85 -o d
```
where the multiscale file, n_msfile, is:

```
1 parent
2 parent
3 parent
4 parent
5 file(custom_orientations)
```

and custom_orientations is:
```
  -0.021953224731    0.020743134319   -0.003282992393
  -0.021276013589    0.012385454484    0.010054691749
   0.022558911077   -0.000949905588   -0.025179887194
  -0.008761448020    0.027661805695   -0.004542095549
  -0.004759898017    0.005199491020   -0.002886854952
```

Visualize:
```
neper -V d.tess -datacellcol "ori" -datacelltrs 0.5 -print img
```
![image](https://user-images.githubusercontent.com/2659376/159433956-e1ea8316-8ffa-4e75-a79c-2b1af4381f11.png)

**Reply by Ming-is** [reply]
Thank you very much. That's what I need.

**Reply by jaja199202** [reply]
Hello Dr @rquey. I  tried to generate 3-scale cell by the following code, but got the error: Unknown argument `vorono'. But 'voronoi' is used in morpho_file. Could you please help me to solve this problem?

`neper -T -n "10::msfile(myfile)::msfile(myfile)" -morpho "gg::voronoi::msfile(morpho_msfile)" -o `

myfile: 
1 1
2 1
3 1
4 1
5 5
6 1
7 1
8 1
9 1
10 1

1::1 1
2::1 1
3::1 1
4::1 1
5::1 from_morpho
5::2 from_morpho
5::3 from_morpho
5::4 from_morpho
5::5 from_morpho
6::1 1
7::1 1
8::1 1
9::1 1
10::1 1
morpho_msfile:
1::1 voronoi
2::1 voronoi
3::1 voronoi
4::1 voronoi
5::1 lamellar(w=0.1)
5::2 lamellar(w=0.1)
5::3 lamellar(w=0.1)
5::4 lamellar(w=0.1)
5::5 lamellar(w=0.1)
6::1 voronoi
7::1 voronoi
8::1 voronoi
9::1 voronoi
10::1 voronoi

**Reply by rquey** [reply]
For me, the command works. So, check your files, command and Neper version (but the version should not be the problem)...

```
$ neper -T -n "10::msfile(myfile)::msfile(myfile)" -morpho "gg::voronoi::msfile(morpho_msfile)" -o foo                                          
                                                    
========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.6.1-3                            
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full) 
Info   : Running on 8 threads.                 
Info   : <https://neper.info>       
Info   : Copyright (C) 2003-2022, and GNU GPL'd, by Romain Quey.        
Info   : Loading initialization file `/home/rquey/.neperrc'...
Info   : ---------------------------------------------------------------
Info   : MODULE  -T loaded with arguments:     
Info   : [ini file] -tesrformat ascii                                                                    
Info   : [com line] -n 10::msfile(myfile)::msfile(myfile) -morpho
         gg::voronoi::msfile(morpho_msfile) -o foo                                                       
Info   : ---------------------------------------------------------------
Info   : Reading input data...                                                                           
Info   : Creating domain...                    
Info   : Creating tessellations...                                                                       
Info   :   [level] Running in multiscale mode (3 scales)...
Info   :   [    1]  1/1 Info   :   - Setting seeds... 100%              
Info   :   - Generating crystal orientations...                                                          
Info   :   - Running tessellation...           
Info   :     > Initial solution: f   =10.585826080
Info   :     > Iteration   5078: fmin=0.112293482 f=0.112293483         
Info   :     > Final solution  : f   =0.112293482 (5078 iterations)
Info   :     > Reached `eps' criterion.
Info   :   [    2]  1/10 Info   :   - Setting seeds...                  
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...
Info   :   [    2]  2/10 Info   :   - Setting seeds...                  
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...
Info   :   [    2]  3/10 Info   :   - Setting seeds...                  
Info   :   - Generating crystal orientations...                                                          
Info   :   - Running tessellation...                                                                     
Info   :   [    2]  4/10 Info   :   - Setting seeds...                   
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...       
Info   :   [    2]  5/10 Info   :   - Setting seeds...                  
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...                                                                     
Info   :   [    2]  6/10 Info   :   - Setting seeds...                  
Info   :   - Generating crystal orientations...                                                          
Info   :   - Running tessellation...                                                                     
Info   :   [    2]  7/10 Info   :   - Setting seeds...                  
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...
Info   :   [    2]  8/10 Info   :   - Setting seeds...                                                   
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...                                                                     
Info   :   [    2]  9/10 Info   :   - Setting seeds...                                                   
Info   :   - Generating crystal orientations...                                                          
Info   :   - Running tessellation...                                                                     
Info   :   [    2]  10/10 Info   :   - Setting seeds...                                                  
Info   :   - Generating crystal orientations...                                                          
Info   :   - Running tessellation...                                                                     
Info   :   [    3]  1/14 Info   :   - Setting seeds...                                                   
Info   :   - Generating crystal orientations...                                                          
Info   :   - Running tessellation...                
Info   :   [    3]  2/14 Info   :   - Setting seeds...                                                   
Info   :   - Generating crystal orientations...     
Info   :   - Running tessellation...                
Info   :   [    3]  3/14 Info   :   - Setting seeds...                  
Info   :   - Generating crystal orientations...                                                          
Info   :   - Running tessellation...                
Info   :   [    3]  4/14 Info   :   - Setting seeds...                  
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...
Info   :   [    3]  5/14 Info   :   - Running tessellation...           
Info   :   - Generating crystal orientations...                                                          
Info   :   [    3]  6/14 Info   :   - Running tessellation...           
Info   :   - Generating crystal orientations...
Info   :   [    3]  7/14 Info   :   - Running tessellation...                                            
Info   :   - Generating crystal orientations...                                                          
Info   :   [    3]  8/14 Info   :   - Running tessellation...                                            
Info   :   - Generating crystal orientations...                                                          
Info   :   [    3]  9/14 Info   :   - Running tessellation...                                            
Info   :   - Generating crystal orientations...
Info   :   [    3]  10/14 Info   :   - Setting seeds...                                                  
Info   :   - Generating crystal orientations...                                                          
Info   :   - Running tessellation...                                                                     
Info   :   [    3]  11/14 Info   :   - Setting seeds...                                                  
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...              
Info   :   [    3]  12/14 Info   :   - Setting seeds...                 
Info   :   - Generating crystal orientations...                                                          
Info   :   - Running tessellation...   
Info   :   [    3]  13/14 Info   :   - Setting seeds...                 
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...
Info   :   [    3]  14/14 Info   :   - Setting seeds...                 
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...
Info   : Flattening 3-scale tessellation...                                                              
Info   :   [level] Flattening 3-scale tessellation...                                                    
Info   :   [    2]  10/10: id=11                                                                         
Info   :   [    1]  1/1: id=1                                            
Info   : Writing results...                    
Info   :     [o] Writing file `foo.tess'...
Info   :     [o] Wrote file `foo.tess'.                                                                  
Info   : Elapsed time: 4.594 secs.             
========================================================================   
```

**Reply by smallrain1996** [reply]
> For me, the command works. So, check your files, command and Neper version (but the version should not be the problem)...
> 
> ```
> $ neper -T -n "10::msfile(myfile)::msfile(myfile)" -morpho "gg::voronoi::msfile(morpho_msfile)" -o foo                                          
>                                                     
> ========================    N   e   p   e   r    =======================
> Info   : A software package for polycrystal generation and meshing.
> Info   : Version 4.6.1-3                            
> Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full) 
> Info   : Running on 8 threads.                 
> Info   : <https://neper.info>       
> Info   : Copyright (C) 2003-2022, and GNU GPL'd, by Romain Quey.        
> Info   : Loading initialization file `/home/rquey/.neperrc'...
> Info   : ---------------------------------------------------------------
> Info   : MODULE  -T loaded with arguments:     
> Info   : [ini file] -tesrformat ascii                                                                    
> Info   : [com line] -n 10::msfile(myfile)::msfile(myfile) -morpho
>          gg::voronoi::msfile(morpho_msfile) -o foo                                                       
> Info   : ---------------------------------------------------------------
> Info   : Reading input data...                                                                           
> Info   : Creating domain...                    
> Info   : Creating tessellations...                                                                       
> Info   :   [level] Running in multiscale mode (3 scales)...
> Info   :   [    1]  1/1 Info   :   - Setting seeds... 100%              
> Info   :   - Generating crystal orientations...                                                          
> Info   :   - Running tessellation...           
> Info   :     > Initial solution: f   =10.585826080
> Info   :     > Iteration   5078: fmin=0.112293482 f=0.112293483         
> Info   :     > Final solution  : f   =0.112293482 (5078 iterations)
> Info   :     > Reached `eps' criterion.
> Info   :   [    2]  1/10 Info   :   - Setting seeds...                  
> Info   :   - Generating crystal orientations...
> Info   :   - Running tessellation...
> Info   :   [    2]  2/10 Info   :   - Setting seeds...                  
> Info   :   - Generating crystal orientations...
> Info   :   - Running tessellation...
> Info   :   [    2]  3/10 Info   :   - Setting seeds...                  
> Info   :   - Generating crystal orientations...                                                          
> Info   :   - Running tessellation...                                                                     
> Info   :   [    2]  4/10 Info   :   - Setting seeds...                   
> Info   :   - Generating crystal orientations...
> Info   :   - Running tessellation...       
> Info   :   [    2]  5/10 Info   :   - Setting seeds...                  
> Info   :   - Generating crystal orientations...
> Info   :   - Running tessellation...                                                                     
> Info   :   [    2]  6/10 Info   :   - Setting seeds...                  
> Info   :   - Generating crystal orientations...                                                          
> Info   :   - Running tessellation...                                                                     
> Info   :   [    2]  7/10 Info   :   - Setting seeds...                  
> Info   :   - Generating crystal orientations...
> Info   :   - Running tessellation...
> Info   :   [    2]  8/10 Info   :   - Setting seeds...                                                   
> Info   :   - Generating crystal orientations...
> Info   :   - Running tessellation...                                                                     
> Info   :   [    2]  9/10 Info   :   - Setting seeds...                                                   
> Info   :   - Generating crystal orientations...                                                          
> Info   :   - Running tessellation...                                                                     
> Info   :   [    2]  10/10 Info   :   - Setting seeds...                                                  
> Info   :   - Generating crystal orientations...                                                          
> Info   :   - Running tessellation...                                                                     
> Info   :   [    3]  1/14 Info   :   - Setting seeds...                                                   
> Info   :   - Generating crystal orientations...                                                          
> Info   :   - Running tessellation...                
> Info   :   [    3]  2/14 Info   :   - Setting seeds...                                                   
> Info   :   - Generating crystal orientations...     
> Info   :   - Running tessellation...                
> Info   :   [    3]  3/14 Info   :   - Setting seeds...                  
> Info   :   - Generating crystal orientations...                                                          
> Info   :   - Running tessellation...                
> Info   :   [    3]  4/14 Info   :   - Setting seeds...                  
> Info   :   - Generating crystal orientations...
> Info   :   - Running tessellation...
> Info   :   [    3]  5/14 Info   :   - Running tessellation...           
> Info   :   - Generating crystal orientations...                                                          
> Info   :   [    3]  6/14 Info   :   - Running tessellation...           
> Info   :   - Generating crystal orientations...
> Info   :   [    3]  7/14 Info   :   - Running tessellation...                                            
> Info   :   - Generating crystal orientations...                                                          
> Info   :   [    3]  8/14 Info   :   - Running tessellation...                                            
> Info   :   - Generating crystal orientations...                                                          
> Info   :   [    3]  9/14 Info   :   - Running tessellation...                                            
> Info   :   - Generating crystal orientations...
> Info   :   [    3]  10/14 Info   :   - Setting seeds...                                                  
> Info   :   - Generating crystal orientations...                                                          
> Info   :   - Running tessellation...                                                                     
> Info   :   [    3]  11/14 Info   :   - Setting seeds...                                                  
> Info   :   - Generating crystal orientations...
> Info   :   - Running tessellation...              
> Info   :   [    3]  12/14 Info   :   - Setting seeds...                 
> Info   :   - Generating crystal orientations...                                                          
> Info   :   - Running tessellation...   
> Info   :   [    3]  13/14 Info   :   - Setting seeds...                 
> Info   :   - Generating crystal orientations...
> Info   :   - Running tessellation...
> Info   :   [    3]  14/14 Info   :   - Setting seeds...                 
> Info   :   - Generating crystal orientations...
> Info   :   - Running tessellation...
> Info   : Flattening 3-scale tessellation...                                                              
> Info   :   [level] Flattening 3-scale tessellation...                                                    
> Info   :   [    2]  10/10: id=11                                                                         
> Info   :   [    1]  1/1: id=1                                            
> Info   : Writing results...                    
> Info   :     [o] Writing file `foo.tess'...
> Info   :     [o] Wrote file `foo.tess'.                                                                  
> Info   : Elapsed time: 4.594 secs.             
> ========================================================================   
> ```

@jaja199202 @rquey The reason should be that the last line of myfile and morpho_msfile does not have a carriage return

**Reply by jaja199202** [reply]
Yes, it is. Thanks a lot for your reply.

**Reply by wanli096** [reply]
Dear Romain,
May I ask if it is possible to determine the multi-scale cell embedding through the ID? For example, the first-level embedding can be divided into ID1 and ID2 based on the X-coordinate. Then the second-level embedding can be done for ID2, while ID1 does not undergo the second-level embedding.

**Reply by rquey** [reply]
Use `-statcell "scaleid(1),scaleid(2)"` to know the "parent" cells of a cell at scale 1, 2.  Works for any number of scales...

**Reply by wanli096** [reply]
Thank you for your reply!

I used the above command to generate a file containing ID information, but I want to distinguish secondary tessellation by ID, where secondary tessellation is only performed in some units. Can this be achieved? For example, the ID of units 1-7 is 1, the ID of units 8 is 2, and only units 8 are divided during secondary tessellation. May I ask if this can be achieved through instructions?

![image](https://github.com/neperfepx/neper/assets/109226361/d9a1a7ac-583c-4d84-a40e-dbddfb52373e)

**Reply by rquey** [reply]
In the new version (4.6.2-3), use `-statcell scale`:

```
neper -T -n "10::msfile(nb)" -statcell scale -dim 2 -id 2 -o d
neper -V d.tess -datacellcol scale -print img
```

nb:
```
1  1
2  1
3  1
4  1
5  1
6  1
7 10
8 10
9 10
10 10
```

![img](https://github.com/neperfepx/neper/assets/2659376/cbeaf3a9-6806-4637-b642-3273ffee895b)

**Reply by 86zhangzhongwen** [reply]
> 1:
> 
> 您必须使用[多尺度单元格文件](https://neper.info/doc/fileformat.html#multiscale-cell-file)将不同的选项值传递给每个第一尺度单元格：
> 
> ```
> neper -T -n "10::msfile(n_msfile)" -morpho "gg::msfile(morpho_msfile)" -o d
> ```
> 
> 其中，多尺度文件为：
> 
> n_msfile：
> 
> ```
> 1 1
> 2 1
> 3 1
> 4 1
> 5 from_morpho
> 6 1
> 7 1
> 8 1
> 9 1
> 10 1
> ```
> 
> morpho_msfile：
> 
> ```
> 1 voronoi
> 2 voronoi
> 3 voronoi
> 4 voronoi
> 5 lamellar(w=0.1)
> 6 voronoi
> 7 voronoi
> 8 voronoi
> 9 voronoi
> 10 voronoi
> ```
> 
> 可视 化：
> 
> ```
> neper -V d.tess -datacellcol "scaleid(1)" -datacelltrs 0.5 -print img
> ```
> 
> ![图像](https://user-images.githubusercontent.com/2659376/157269047-ef0f0208-c101-4e00-9f39-1bee1ebe4b8e.png)
> 
> 2:
> 
> Same principle:
> 
> ```
> neper -T -n "10::msfile(n_msfile2)::msfile(n_msfile3)" -morpho "gg::voronoi::msfile(morpho_msfile3)" -o d
> ```
> 
> where the multiscale files are:
> 
> n_msfile2:
> 
> ```
> 1 1
> 2 1
> 3 1
> 4 1
> 5 5
> 6 1
> 7 1
> 8 1
> 9 1
> 10 1
> ```
> 
> n_msfile3:
> 
> ```
> 1::1 1
> 2::1 1
> 3::1 1
> 4::1 1
> 5::1 from_morpho
> 5::2 from_morpho
> 5::3 from_morpho
> 5::4 from_morpho
> 5::5 from_morpho
> 6::1 1
> 7::1 1
> 8::1 1
> 9::1 1
> 10::1 1
> ```
> 
> morpho_msfile3:
> 
> ```
> 1::1 voronoi
> 2::1 voronoi
> 3::1 voronoi
> 4::1 voronoi
> 5::1 lamellar(w=0.1)
> 5::2 lamellar(w=0.1)
> 5::3 lamellar(w=0.1)
> 5::4 lamellar(w=0.1)
> 5::5 lamellar(w=0.1)
> 6::1 voronoi
> 7::1 voronoi
> 8::1 voronoi
> 9::1 voronoi
> 10::1 voronoi
> ```
> 
> Visualize:
> 
> ```
> neper -V d.tess -datacellcol "scaleid(1)" -datacelltrs 0.5 -print img
> ```
> 
> ![image](https://user-images.githubusercontent.com/2659376/157270057-c52638a1-5bf7-43c5-abb0-016e0157e66c.png)
> 
> It is also possible to merge the different multiscale cell files ( and , in the above) - Neper scans the file until it finds the correct multiscale id at the beginning of the line.`n_msfile2``n_msfile3`


![微信图片_20241104150957](https://github.com/user-attachments/assets/9c8eb15c-7f0e-4d29-9eab-423ba278dd45)
How to set the layered structure to have different widths, such as the layered structures with widths of 0.9 and 0.1 appearing alternately in the picture

**Reply by rquey** [reply]
Use [`-morpho lamellar(w=0.09:0.01)`](https://neper.info/doc/neper_t.html#cmdoption-morpho).

**Comment by charanjo** [comment]
Is it possible to get three different Voronoi grain size across a rectangular domain?

**Reply by rquey** [reply]
If you mean a 3-scale tessellation, yes:

```
neper -T -n 10::10::10 -o 3scale
neper -V 3scale.tess -datacellcol "scaleid(1)" -print img1
neper -V 3scale.tess -datacellcol "scaleid(2)" -print img2
```

![img](https://github.com/neperfepx/neper/assets/2659376/44e56557-6f81-4f54-9eb3-3fed999787eb)
![img2](https://github.com/neperfepx/neper/assets/2659376/60bd5dfc-85dc-4304-ab2d-4b73726bdeb5)

**Reply by charanjo** [reply]
Hi,
Not a 3 -Scale tessellation. For example, in a 2D domain, I need a grain size of say 10 15 and 20 at different depths. Is it possible to obtain that?

**Reply by wanli096** [reply]
Dear Romain,
May I ask if it is possible to determine the multi-scale cell embedding through the ID? For example, the first-level embedding can be divided into ID1 and ID2 based on the X-coordinate. Then the second-level embedding can be done for ID2, while ID1 does not undergo the second-level embedding.
